ys001-yggdrasil-core-specification.md

```diff
@@ -0,0 +1,861 @@
+completed and the node **should not** send any additional search requests
+related to this search. If not, then the node which handled the request **must**
+be added to a list of visited nodes, so that it can be ignored if included in
+any future responses. Any information about closer nodes, included in the response,
+**must** be checked to confirm that they are closer to the target Node ID than the
+request handler. Any nodes which satisfy this condition and do not appear on the
+list of nodes already visited in this search **should** be added to a list of
+known nodes to contact in later search iterations.
+
+After any response from a node which is not the target of a search, or after a
+timeout in the event of no response, the node **must** select the node from the
+list of unchecked destinations which is closest to the target Node ID, send a
+request to that node, and remove the node from the list. This process **must**
+repeat iteratively until either the a node matching the known bits of the target
+Node ID is contacted, or until the list of nodes to visit is exhaused and a timeout
+passes (to give the last node a chance to respond).
+
+When a search is initialized, the list of visited nodes **must** be empty, and the
+list of nodes to visit in later search iterations **must** be filled by the nodes
+closest to the destination which are already present in the searching node's DHT.
+
+### Caching
+
+In a Chord-like DHT, a node's predecessor or successor on the DHT may not be
+close to the node in metric space. As searches take place, a node may discover
+other nodes which are closer in metric space than the node's predecessor or
+successor.
+
+In order to help to speed up lookups, an Yggdrasil node **may** cache these
+closer nodes as additional DHT neighbours to use during searches. This allows us
+the node to maintain a small database of DHT neighbours that are relatively
+close to the node on the tree.
+
+---
+
+## Forwarding
+
+All nodes in an Yggdrasil network have the capacity to forward traffic on behalf
+of other network nodes. This happens if:
+
+1. The node has peering connections to more than one node
+1. The current node is on an optimal path between two other network nodes which
+   are exchanging traffic
+
+### Next-hop selection
+
+When a message is received for forwarding, the node **must** make a decision as
+to which node to forward onto next, except where the destination coordinates of
+a message are equal to the coordinates of the current node, in which case the
+message is considered to have been delivered.
+
+The node should evaluate each direct peering and determine which peering will
+take the message closest to the target destination coordinates. This is done by
+calculating the metric distance between the peer's coordinates and the target
+coordinates, selecting the peering which returns the shortest distance.
+
+A node **must not** forward traffic to any node which is not closer to the
+destination.
```

Comment:
In other words, if it can't find closer node and coordinates do not match, it should drop, right? Maybe could be spelled out explicitly.